src/core/ext/transport/chaotic_good/chaotic_good_transport.h

```diff
@@ -81,58 +98,47 @@ class ChaoticGoodTransport : public RefCounted<ChaoticGoodTransport> {
               << " "
               << (frame_header.ok() ? frame_header->ToString()
                                     : frame_header.status().ToString());
-          // Read header and trailers from control endpoint.
-          // Read message padding and message from data endpoint.
-          return If(
-              frame_header.ok(),
-              [this, &frame_header] {
-                const uint32_t message_padding = frame_header->message_padding;
-                const uint32_t message_length = frame_header->message_length;
-                return Map(
-                    TryJoin<absl::StatusOr>(
-                        control_endpoint_.Read(frame_header->GetFrameLength()),
-                        data_endpoint_.Read(message_length + message_padding)),
-                    [frame_header = *frame_header, message_padding](
-                        absl::StatusOr<std::tuple<SliceBuffer, SliceBuffer>>
-                            buffers)
-                        -> absl::StatusOr<std::tuple<FrameHeader, BufferPair>> {
-                      if (!buffers.ok()) return buffers.status();
-                      SliceBuffer data_read = std::move(std::get<1>(*buffers));
-                      if (message_padding > 0) {
-                        data_read.RemoveLastNBytesNoInline(message_padding);
-                      }
-                      return std::tuple<FrameHeader, BufferPair>(
-                          frame_header,
-                          BufferPair{std::move(std::get<0>(*buffers)),
-                                     std::move(data_read)});
-                    });
-              },
-              [&frame_header]() {
-                return [status = frame_header.status()]() mutable
-                       -> absl::StatusOr<std::tuple<FrameHeader, BufferPair>> {
-                  return std::move(status);
-                };
-              });
+          return frame_header;
+        },
+        [this](FrameHeader frame_header) {
+          current_frame_header_ = frame_header;
+          auto con = frame_header.payload_connection_id == 0
+                         ? &control_endpoint_
+                         : &data_endpoint_;
+          return con->Read(frame_header.payload_length +
+                           frame_header.Padding(decode_alignment_));
+        },
+        [this](SliceBuffer payload)
+            -> absl::StatusOr<std::tuple<FrameHeader, SliceBuffer>> {
+          payload.RemoveLastNBytes(
```

Comment:
I think if padding > 0, you'll have to use RemoveLastNBytesNoInline to avoid creating and passing inlined slices

Reply:
done